core/src/local_connection.rs

```diff
@@ -1,82 +1,239 @@
+        }
+    }
+
+    pub fn run_method(
+        &self,
+        context: &mut UpdateContext<'_, 'gc>,
+        method_name: AvmString<'gc>,
+        arguments: Vec<AmfValue>,
+    ) {
+        match self {
+            LocalConnectionKind::Avm2(domain, object) => {
+                object.run_method(context, *domain, method_name, arguments);
+            }
+            LocalConnectionKind::Avm1(object) => {
+                if let Err(e) =
+                    Avm1LocalConnectionObject::run_method(context, *object, method_name, arguments)
+                {
+                    tracing::error!("Unhandled AVM1 error during LocalConnection onStatus: {e}");
+                }
+            }
+        }
     }
 }
 
 #[derive(Collect)]
 #[collect(no_drop)]
-pub struct LocalConnection<'gc> {
-    object: LocalConnectionKind<'gc>,
+pub struct QueuedMessage<'gc> {
+    source: LocalConnectionKind<'gc>,
+    kind: QueuedMessageKind<'gc>,
+}
 
-    connection_name: AvmString<'gc>,
+#[derive(Collect)]
+#[collect(no_drop)]
+pub enum QueuedMessageKind<'gc> {
+    Failure,
+    Message {
+        #[collect(require_static)]
+        connection_name: WString,
+        method_name: AvmString<'gc>,
+        #[collect(require_static)]
+        arguments: Vec<AmfValue>,
+    },
 }
 
-impl<'gc> LocalConnection<'gc> {
-    pub fn new(
-        object: impl Into<LocalConnectionKind<'gc>>,
-        connection_name: AvmString<'gc>,
-    ) -> Self {
-        Self {
-            object: object.into(),
-            connection_name,
+impl<'gc> QueuedMessageKind<'gc> {
+    pub fn deliver(self, source: LocalConnectionKind<'gc>, context: &mut UpdateContext<'_, 'gc>) {
+        match self {
+            QueuedMessageKind::Failure => {
+                source.send_status(context, "error");
+            }
+            QueuedMessageKind::Message {
+                connection_name,
+                method_name,
+                arguments,
+            } => {
+                if let Some(receiver) = context.local_connections.find_listener(&connection_name) {
+                    source.send_status(context, "status");
+                    receiver.run_method(context, method_name, arguments);
+                } else {
+                    source.send_status(context, "error");
+                }
+            }
         }
     }
 }
 
+/// An opaque handle to an actively listening LocalConnection.
+/// Owning this handle represents ownership of a LocalConnection;
+/// However, a LocalConnection must be manually closed, you can't just Drop this handle.
+#[derive(Debug)]
+pub struct LocalConnectionHandle(WString);
+
 /// Manages the collection of local connections.
 pub struct LocalConnections<'gc> {
-    connections: SlotMap<LocalConnectionHandle, LocalConnection<'gc>>,
+    connections: FnvHashMap<WString, LocalConnectionKind<'gc>>,
+    messages: Vec<QueuedMessage<'gc>>,
 }
 
-unsafe impl<'gc> Collect for LocalConnections<'gc> {
+unsafe impl Collect for LocalConnections<'_> {
```

Comment:
This should be able to `derive(Collect)` now

Reply:
Can't, WString doesn't impl Collect :(